### ISSUE TITLE
Continents Backwards Compatibility Test Fails

### ISSUE THREAD

**Issue by JuxhinDB** [JuxhinDB/gw2-api-interface#10]
Commit: **ddf369bec070bb4d1639e07c50476f86ee3a56e8**

---

The following unittest seems to be failing:

https://github.com/JuxhinDB/gw2-api-interface/blob/ddf369bec070bb4d1639e07c50476f86ee3a56e8/test/test_continents.py#L316-L339

Specifically the last one. It appears that for some reason, the Mock Adapter is returning `[1, 2]` instead of the actual response. After some troubleshooting, I tested this locally and the API response seems to be pulled correctly:

```
>>> api
<GuildWars2Client https://api.guildwars2.com
Version: v2
API Key: None
Language: en
Proxy: None
Verify SSL?: True>
>>> api.build.get()
{'id': 90153}
>>> api.continents.get()
[1, 2]
>>> api.continents.get(ids=[1, 2])
[{'name': 'Tyria', 'continent_dims': [49152, 49152], 'floors': [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, -2, -3, -4, -5, -6, -7, -8, -9, -10, -11, -12, -13, -14, -15, -16, -18, -19, -20, -21, -22, 20, 21, 22, 23, 24, 25, 28, 29, 39, 42, 26, 27, 30, 34, 36, 37, 38, 40, 41, 43, 44, 45, 46, 47, 48, 49, -24, -26, -27, -28, -29, -30, -31, -33, -34, -35, -36, -32, 50, -37, 53], 'max_zoom': 7, 'id': 1, 'min_zoom': 0}, {'name': 'Mists', 'continent_dims': [16384, 16384], 'floors': [1, 3, 5, 6, 7, 8, 9, 10, 12, 13, 14, 15, 16, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, -27, -28, -29, -30, -31, 35, 32, -32, -33, 37, 39, 40], 'max_zoom': 6, 'id': 2, 'min_zoom': 0}]
>>>
```

**Comment by JuxhinDB** [comment]
@shoeffner -- Would you be able to take a quick look at this? I ran it through a debugger and for some reason, during the unittest, calling:

`gw2_client.continents.get(ids=[1, 2])` is always returning `[1, 2]`. However when I call it outside of the unittest environment, I'm getting the correct response.

**Comment by shoeffner** [comment]
I will take a look!

Update 1:
Running locally it always passes (using the current master, ddf369b). 
```
$ pytest -v
...
platform darwin -- Python 3.6.5, pytest-3.6.1, py-1.5.3, pluggy-0.6.0 -- /Users/shoeffner/.local/share/virtualenvs/gw2-api-interface-0IQaR2wa/bin/python3.6
cachedir: .pytest_cache
rootdir: /Users/shoeffner/Projects/gw2-api-interface, inifile:
plugins: requests-mock-1.5.0
collected 9 items                                                                                                                                                                          
...
test/test_continents.py::test_continents_backwards_compatibility PASSED
...
9 passed in 0.14 seconds
```

I will try to get it to fail. I remember having had that issue, but I thought I had it resolved at some point. Which Python version are you using?

**Comment by bmhenry** [comment]
@shoeffner The CI build was passing just fine for 3.6 but failing for 3.4 and 3.5.

```
=================================== FAILURES ===================================
___________________ test_continents_backwards_compatibility ____________________
gw2_client = <GuildWars2Client https://api.guildwars2.com
Version: v2
API Key: None
Language: en
Proxy: None
Verify SSL?: True>
mock_adapter = <requests_mock.adapter.Adapter object at 0x7f8dd486a668>
    def test_continents_backwards_compatibility(gw2_client, mock_adapter):
        """This test assures that the new changes to introduce all sub parts of
        the continents API does not break the old behaviour, that is accessing
        the continents via id= and ids=.
    
        Args:
            gw2_client: The pytest "gw2_client" fixture.
            mock_adapter: The pytest "mock_adapter" fixture.
        """
        url_to_file = {
            'continents': 'continents',
            'continents/2': 'continents2',
            'continents?ids=1,2': 'continents1_2',
        }
        register_urls_to_files(mock_adapter, url_to_file)
    
        expected_continent_list = load_mock_json('continents')
        assert gw2_client.continents.get() == expected_continent_list, '/continents failed'
    
        expected_single_id = load_mock_json('continents2')
        assert gw2_client.continents.get(id=2) == expected_single_id, '/continents?id=2 failed'
    
        expected_multi_id = load_mock_json('continents1_2')
>       assert gw2_client.continents.get(ids=[1, 2]) == expected_multi_id, '/continents?ids=1,2 failed'
E       AssertionError: /continents?ids=1,2 failed
E       assert [1, 2] == [{'continent_dims': [49152...': 2, 'max_zoom': 6, ...}]
E         At index 0 diff: 1 != {'name': 'Tyria', 'continent_dims': [49152, 49152], 'floors': [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, -2, -3, -4, -5, -6, -7, -8, -9, -10, -11, -12, -13, -14, -15, -16, -18, -19, -20, -21, -22, 20, 21, 22, 23, 24, 25, 28, 29, 39, 42, 26, 27, 30, 34, 36, 37, 38, 40, 41, 43, 44, 45, 46, 47, 48, 49, -24, -26, -27, -28, -29, -30, -31, -33, -34, -35, -36, -32, 50, -37], 'max_zoom': 7, 'id': 1, 'min_zoom': 0}
E         Use -v to get the full diff
test/test_continents.py:339: AssertionError
====================== 1 failed, 9 passed in 0.26 seconds ======================
```

**Comment by JuxhinDB** [comment]
@shoeffner - I was able to replicate it locally consistently on Python 3.6 in a virtualenv. This only occurs when running the test suite; I tried debugging it but everything seemed to be in order. However when running `gw2_client.continents.get(ids=[1, 2])` during the test run, the result is always `[1, 2]` which I found weird. 

The only mock response that has that, is `continents.json` which made me think that the Mock Adapter for some reason was returning that. It could also be some odd race condition where another test is adjusting the `url_to_filename` before or after this faulty test, which could explain the intermittent issues. Could you try running the test case alone (not as part of the entire test suite) and see if you can reproduce it?

**Comment by shoeffner** [comment]
I will try, thank your for singling it out!

**Comment by shoeffner** [comment]
I tried it locally with 3.4, 3.5, 3.6 and 3.7, and all cases work fine. I split the failing test into it's own small unit test and will let travis build it... See #12.

edit: To clarify, I used `pytest -k compatibility_ids` to run the test, did you try another way?

**Comment by shoeffner** [comment]
So if you are right and it is/was a race condition (which I can't test nor really understand why it would be one, as url_to_filename should only be additions... hmhm), #12 seems to have this resolved. Maybe you can kick off travis once or two more times to assure it is working? I really can't reproduce it locally.

**Comment by JuxhinDB** [comment]
@shoeffner Do not worry, I thank you for looking into this. I do it from PyCharm which internally runs the same command more or less. I'm restarting the build 5 or so times and will see if they all pass and will report back on the PR :-)!

**Comment by JuxhinDB** [comment]
Fixed as of PR #12 is seems. Closing!
